### PR TITLE
feat!: redesign website

### LINKED ISSUE #311: migrate website content to `@nuxt/content`
right now the content of website is coded in the pages (getting data from store), I suggest that we use [@nuxt/content](https://github.com/nuxt/content), this will add support for MarkDown which will be more convenient to use and manage.

I would be happy to work on this if approved.

### PR DESCRIPTION
**Pull request by arashsheyda**
## What issue is this referencing?
this PR will closes #311, #139 (with breaking changes)

## Description
- use `pnpm` instead of `npm` (as it is much faster)
- latest `nuxt` with `@nuxt/devtools` support
- `@nuxt/content` for writing static content in markdwon
- `@nuxt/eslint` for eslint setup with new flat configuration
- `@nuxtjs/seo` for SEO tools like `ogImage`, `SchemaOrg`, `LinkChecker` and ...
- new component structure
- `nuxt-icon` for iconify icons
- moved the data from `store` to `constants` (with auto-import enabled now). maybe we can move it to `composables` in the future.

- improved lighthouse score: (the score is improved but the performance might not always be at 100, jsut a little heads up)
Screenshot 2024-03-22 at 9 21 43 PM

### CODE REVIEW

File: constants/slack.ts
Comment (MandyMeindersma) [inline constants/slack.ts:14]: to showcase local events and meetups about tech

File: constants/slack.ts
Comment (MandyMeindersma) [inline constants/slack.ts:17]: nit pick, maybe move random to the bottom

**Comment by MandyMeindersma** [comment]
the board recruitment page seems to 404 https://deploy-preview-315--dev-edmonton.netlify.app/board_recruiting

File: pages/[...slug].vue
Comment (MandyMeindersma) [inline pages/[...slug].vue:51]: we did have a fun 404 template

File: pages/[...slug].vue
Comment (MandyMeindersma) [inline pages/[...slug].vue:21]: nit: description could be -  Fostering a stronger software development community in Edmonton. We are made up of people who believe that an impactful, diverse, and collaborative developer community enriches us all. We realize this vision by sharing knowledge and ideas, understanding the community and its needs, promoting events and initiatives, and supporting technology meetups and organizations.

File: pages/[...slug].vue
Comment (MandyMeindersma) [inline pages/[...slug].vue:4]: Where does it decide what to match? Like anything in the content folder? but them how does meetups work?

File: public/assets/images/edmonton.jpg
Comment (MandyMeindersma) [inline public/assets/images/edmonton.jpg:1]: You have the same image `public/edmonton-fit-org.jpg`

do we need both?

File: server/api/contact.post.ts
Comment (MandyMeindersma) [inline server/api/contact.post.ts:1]: I assume we will wait for this to merge?

File: pages/[...slug].vue
Comment (MandyMeindersma) [inline pages/[...slug].vue:13]: This one I am not sure what that means, or how to fix it so we might also have to wait for this to merge?

**Comment by MandyMeindersma** [comment]
Okay general thoughts:

1. This is beautiful and I am so excited. Thank you for all the work you have put into it. 
2. How do we deal with the TODO's? Is your plan to make issues for them and fix later or fix before the pr merges in?
3. Most important question: when I am confused about something with nuxt content in the next 6 months can I DM you hahahah

File: app.vue
Comment (MandyMeindersma) [inline app.vue:5]: is the SEO description supposed to be shorter? Like 140-160 characters?

I assume this would be too log:

Fostering a stronger software development community in Edmonton. We are made up of people who believe that an impactful, diverse, and collaborative developer community enriches us all. We realize this vision by sharing knowledge and ideas, understanding the community and its needs, promoting events and initiatives, and supporting technology meetups and organizations.

File: components/app/ColorMode.vue
Comment (MandyMeindersma) [inline components/app/ColorMode.vue:5]: delete?

File: .env.example
Comment (MandyMeindersma) [inline .env.example:1]: We have the API keys referencing the env in another file? What does this one do?

**Comment by MandyMeindersma** [comment]
@MarkBennett You do not have to review but I think you would find this pr really fun. The resdesign is so cool!

**Comment by deejayjay** [comment]
> Okay general thoughts:
> 
> 1. This is beautiful and I am so excited. Thank you for all the work you have put into it. 
> 2. How do we deal with the TODO's? Is your plan to make issues for them and fix later or fix before the pr merges in?
> 3. Most important question: when I am confused about something with nuxt content in the next 6 months, can I DM you hahahah

@arashsheyda: I really love how good the redesigned landing page looks. Great job on the redesign and the improved light house scores.

@MandyMeindersma: I was literally thinking the same thing you mentioned in #3. It would be nice to have some support if we end up having issues later on. Also, an FYI, I am helping my brother move today, so I can't review the PR until later this evening.

**Comment by MarkBennett** [comment]
> @MarkBennett You do not have to review but I think you would find this pr really fun. The resdesign is so cool!

Thanks for the heads up @MandyMeindersma! I'm just heading out for a spring break trip with the family but I might lurk on this just to see. Very cool and great work @arashsheyda! 😄

**Comment by MarkBennett** [comment]
It has a dark theme! So cool @arashsheyda! 🤯

File: pages/[...slug].vue
Comment (arashsheyda) [reply pages/[...slug].vue:51]: yes, we have to create a `Error` component and use that, that's on the TODO list

File: pages/[...slug].vue
Comment (arashsheyda) [reply pages/[...slug].vue:4]: @MandyMeindersma nuxt content offers some composables, one of them is `queryContent`,

```ts
const { data: page } = await useAsyncData(route.path, () => queryContent(route.path).findOne())
```

here we are getting the current content base on `route.path`, and if not found we ar showing an `404`

File: server/api/contact.post.ts
Comment (arashsheyda) [reply server/api/contact.post.ts:1]: yes, I wasn't sure how you manage contact form

File: pages/[...slug].vue
Comment (arashsheyda) [reply pages/[...slug].vue:13]: so right now if we try to get footer's height and then minus it from screen height, it gives some warnings (in dev only) because the footer instance hasn't rendered yet (in SSR) and it renders in client side (not a big deal for now but we can fix these todo's in the future)

File: app.vue
Comment (arashsheyda) [reply app.vue:5]: yes actually, the ogImage look like this: https://devedmonton.vercel.app/__og-image__/image/og.png

it would be nicer if we could shorten the description a bit :)

Comment (arashsheyda) [reply app.vue:5]: this one is the fallback description, if the page doesn't have a desc. this would show up

File: components/app/ColorMode.vue
Comment (arashsheyda) [reply components/app/ColorMode.vue:5]: maybe not? so in the future we can add coffee mode and etc. ?

File: .env.example
Comment (arashsheyda) [reply .env.example:1]: yes, this one is for contributeurs so they know what variables we have

**Comment by arashsheyda** [comment]
@MandyMeindersma 

> 2. How do we deal with the TODO's? Is your plan to make issues for them and fix later or fix before the pr merges in?

yes, the TODO's are more like nice to have things in the future, except contact page


> 3. Most important question: when I am confused about something with nuxt content in the next 6 months can I DM you hahahah

yes for sure, anytime! I'de be super happy to help with anything related to vue & nuxt :DD

**Comment by arashsheyda** [comment]
@MarkBennett @deejayjay thanks!

File: pages/[...slug].vue
Comment (arashsheyda) [reply pages/[...slug].vue]: update:

Screenshot 2024-03-23 at 9 45 03 AM

**Comment by DerrykBoyd** [comment]
I'm not going to go through a whole review due to the quantity of changes, but I do have a few comments after a quick glance.

Great work on these changes by the way.

> use pnpm instead of npm (as it is much faster)

I don't necessarily agree with this change. At the very least I think it should be separated into it's own PR.

- It looks like our `test` and `format` npm commands have been removed, and also removed from the PR guidelines.
- It looks like our test setup has been removed entirely. That's a pretty big regression.

I could maybe do a more in depth look later, but these are the things that jumped out for me right away that I thought might merit some discussion.

**Comment by arashsheyda** [comment]
> I don't necessarily agree with this change: `pnpm`

here is why we sohuld use `pnpm` [pnpm vs npm](https://pnpm.io/pnpm-vs-npm)
also in my experience `pnpm` works better in many ways but if you want we can roll back to `npm` 😄

> At the very least I think it should be separated into it's own PR.

I thought including it here made sense because of all the changes. But if you think it's better separate, we can do that 

> It looks like our `test` and `format` npm commands have been removed, and also removed from the PR guidelines.

`format` comamnd has been replace with `lint` (prettier is good but sometimes it can be too strict, and I think we can have what prettier has with eslint)
also there is [this article about eslint & prettier](https://antfu.me/posts/why-not-prettier)


> It looks like our test setup has been removed entirely. That's a pretty big regression.

the `tests` were removed as it had 3 basic tests, and I thought starting fresh will help us build better tests later on.

> I could maybe do a more in depth look later, but these are the things that jumped out for me right away that I thought might merit some discussion.

thanks @DerrykBoyd

**Comment by DerrykBoyd** [comment]
> I thought including it here made sense because of all the changes.

It's definitely best practice to keep PRs as small as possible and single focus. Otherwise things can get stuck in limbo since now you end up with multiple discussions in the PR trying to work through opinions on each change.

### on pnpm vs npm

That's an opinionated change, my main objection here is it's another small hurdle for contributors. But I'm not totally against it. If this does stay, I would like to see the README updated with clear instructions on how to get started on the project using pnpm, right now it's all npm commands which aren't relevant anymore if this PR is merged.

### formatters vs linters

[Here are the recommendations](https://typescript-eslint.io/troubleshooting/formatting/#problems-with-using-linters-as-formatters) from the `typescript-eslint` project on this topic.

Again, this isn't something I would get behind, but I don't mind if others want to go ahead with it.

### on tests

Yes we didn't have many tests, but what we did have was a working test setup, and that's valuable. Before if any contributor wanted to add a test all they had to do was create a `x.test.js` file and write the test, everything else was handled including running the tests on CI.

Having personally volunteered time to get that setup up and running, it stings a little bit to see it removed without replacement.

Again, overall I don't want these things to detract from the work done here. It's incredible that you volunteered this much time to improve the site! And if this was merged as is, I would still be incredibly pleased. 

Suggested action items before merge for me:

- Update the README so that contributors can quickly get setup with the new tooling
- Bring back a working test setup (optional - maybe a follow up PR)

**Comment by DerrykBoyd** [comment]
@arashsheyda please don't revert the prettier change 😄 I think if others are ok with it it would be good to try the eslint only setup and see how that goes. 

It might help with some issues we had recently with contributors editing files directly in github

**Comment by DerrykBoyd** [comment]
Thanks for doing those other changes! I'm glad the tests are back

Excellent work again @arashsheyda 

Feel free to open up a follow up PR for switching to pnpm

**Comment by DerrykBoyd** [comment]
A couple odd behaviours around scrolling that I noticed in the deploy preview

1. After a navigation there is a jarring jump to the top of the current page before the next page loads (if there is no cache of that page yet)

https://github.com/devedmonton/DES-Website/assets/24216368/ea11307a-7a43-4905-8ab9-f5b9dbaf6256

2. Between around 1020px to 1300px something (I assume the slack animation) is causing the horizontal scroll to trigger and move (see the scrollbar at the bottom)
3. The "As a business" image link appears to be broken (can see in this video as well)

https://github.com/devedmonton/DES-Website/assets/24216368/eb3ab84e-99c7-457d-bd6f-c55f95662436

**Comment by arashsheyda** [comment]
@DerrykBoyd the navigation thing is because we are getting the content of that page from content directory (`@nuxt/content`) and that requires a api fetch, we can make the fetch lazy (so the page would load instantly but we have to show some skeleton UI to user while the data loads) or we can prefetch the data (but it might effect the site's performance) let me know what way you want to approach

2. yes I'm aware of that 😄 I don't have illustrator at the moment but when we convert the illustrations to svg, we can set their size easier

3. that must've been a bad build from netlify (it was working before and it works locally)

**Comment by DerrykBoyd** [comment]
> @DerrykBoyd the navigation thing is because we are getting the content of that page from content directory (`@nuxt/content`) and that requires a api fetch, we can make the fetch lazy (so the page would load instantly but we have to show some skeleton UI to user while the data loads) or we can prefetch the data (but it might effect the site's performance, **edited: the pages are loading faster now with prefetch enabled, the performance is not effected that much**) let me know what way you want to approach

Ideally the solution would be that the scroll on the page doesn't change while waiting for that transition. But I don't think this is a major problem, could be a follow up issue for someone to look into.

> I don't have illustrator at the moment but when we convert the illustrations to svg, we can set their size easier

All good 😄 again, minor issue

>  must've been a bad build from netlify

Possibly, I'm not convinced though. Probably ok to see after final deploy if it's working. What I did notice is that image is not using our image-kit setup while the others are. Also the image in production points directly to unsplash, where the one on the deploy preview seems to be processed by Netlify now.

At this point I'll defer to @MandyMeindersma to get this over the line.

Thanks for you patience with those comments @arashsheyda much appreciated.

File: content/board_recruiting.md
Comment (arashsheyda) [inline content/board_recruiting.md:9]: also I think there are some typos in here, right?

`Compentencies` => `Competencies`
`excercises` => `exercises`
`Visonary` => `Visionary`

Comment (MandyMeindersma) [reply content/board_recruiting.md:9]: omg this is my horrible spelling, please fix if you have time hahaha

**Comment by MandyMeindersma** [comment]
So I think in one of the recent changes we lost the "randomization" of the meetups and community events. 

Like if a refresh the highlighted 5 tiles seem to stay the same?

File: server/api/contact.post.ts
Comment (MandyMeindersma) [reply server/api/contact.post.ts:1]: so I think you can delete this file because I sent a test message on your netlify deploy and it works 🤷‍♀️ 

Must be some netlify magic

image

**Comment by MandyMeindersma** [comment]
I have a couple more comments but they will probably have to wait until Thursday! Sorry to make you wait @arashsheyda 

Thank you again!

**Comment by arashsheyda** [comment]
@MandyMeindersma I have the same code built on vercel and it's working fine (randomization and the unsplash image https://devedmonton.vercel.app/)

**Comment by MandyMeindersma** [comment]
new assets in imagekit:

- https://ik.imagekit.io/des/meetups/DataScience (updated image version)
- https://ik.imagekit.io/des/events/clc.png (updated image version)
- https://ik.imagekit.io/des/events/gdgLogo.png (updated image version)
- https://ik.imagekit.io/des/weeklydevchat.svg?updatedAt=1711891116380
- https://ik.imagekit.io/des/amii.svg?updatedAt=1711891076887
- https://ik.imagekit.io/des/python.svg?updatedAt=1711891265404
- https://ik.imagekit.io/des/edmontonUnlimited.svg?updatedAt=1711891076630
- https://ik.imagekit.io/des/r.svg?updatedAt=1711891284058

File: content/board_recruiting.md
Comment (MandyMeindersma) [inline content/board_recruiting.md:1]: do the other ones all have nubers in front of them because only numbers are rendered in the footer or something?

File: pages/[...slug].vue
Comment (MandyMeindersma) [reply pages/[...slug].vue:4]: ah okay, but where does route path pull from?

Like if I wanted to add a new page, how would I do that?

File: server/api/contact.post.ts
Comment (MandyMeindersma) [reply server/api/contact.post.ts:1]: delete this file

**Comment by MandyMeindersma** [comment]
Again amazing work @arashsheyda 

Other than answering my random nuxt context questions I think there are only 2 things I am worried about:
- Just delete that `server/api/contact.post.ts` file 
- the randomization of the meetups and stuff is still not working on netlify and I would love to see that working but it's not a blocker for this pr

**Comment by arashsheyda** [comment]
@MandyMeindersma the `randomize` function is in `/utils/index.ts`

File: content/board_recruiting.md
Comment (arashsheyda) [reply content/board_recruiting.md:1]: @MandyMeindersma sorry I didn't understand, what do you mean?

Comment (arashsheyda) [reply content/board_recruiting.md:1]: ohh, no the numbers are used for ordering them. if we want to show it in footer we have to add this to the file, like:
```md
---
footer: true
---
```

**Comment by arashsheyda** [comment]
@MandyMeindersma in order for randomize to work we need SSR.

| netlify | vercel |
:--------------:|:-------------:
Screenshot 2024-03-31 at 10 50 09 AM | Screenshot 2024-03-31 at 10 51 00 AM

**Review by MandyMeindersma — APPROVED** [review]

**Comment by MandyMeindersma** [comment]
@all-contributors add @arashsheyda for code and ideas

**Pull request merged**
